synapse/storage/data_stores/main/room.py

```diff
@@ -281,6 +282,149 @@ def is_room_blocked(self, room_id):
+            order_by_asc = True
+        else:
+            raise StoreError(500, "Incorrect value for order_by provided: %s", order_by)
+
+        # Whether to return the list in reverse order
+        if reverse_order:
+            # Flip the boolean
+            order_by_asc = not order_by_asc
+
+        # Create one query for getting the limited number of events that the user asked
+        # for, and another query for getting the total number of events that could be
+        # returned. Thus allowing us to see if there are more events to paginate through
+        info_sql = """
+            SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members
+            FROM room_stats_state state
+            LEFT JOIN room_stats_current curr ON state.room_id = curr.room_id
+            %s
+            ORDER BY %s %s
+            LIMIT ?
+            OFFSET ?
+        """ % (
+            where_statement,
+            order_by_column,
+            "ASC" if order_by_asc else "DESC",
+        )
+
+        # We don't want a LIMIT on the count query. SQLite can't handle an OFFSET
+        # without a LIMIT, however you can hand it `LIMIT -1` to mean "no limit".
+        # Postgres doesn't like this however, and wants you to eliminate the LIMIT
+        # keyword altogether
+        if isinstance(self.database_engine, PostgresEngine):
+            limit_statement = ""
+        else:
+            limit_statement = "LIMIT -1"
+
+        # Use a nested SELECT statement as SQL can't count(*) with an OFFSET
+        count_sql = """
+            SELECT count(*) FROM (
+              SELECT room_id FROM room_stats_state state
+              %s
+              %s
+              OFFSET ?
+            ) AS get_room_ids
+        """ % (
+            where_statement,
+            limit_statement,
+        )
+
+        logger.info("info sql: %s count_sql: %s", info_sql, count_sql)
+
+        def _get_rooms_paginate_txn(txn):
+            # Execute the data query
+            sql_values = (limit, start)
+            if search_term:
+                # Add the search term into the WHERE clause
+                sql_values = (search_term,) + sql_values
+            logging.info("info sql_values: %s", sql_values)
+            txn.execute(info_sql, sql_values)
+            rows = [row for row in txn]
+
+            # Execute the count query
+            sql_values = (start,)
+            if search_term:
+                # Add the search term into the WHERE clause
+                sql_values = (search_term,) + sql_values
+            logging.info("count sql_values: %s", sql_values)
+            txn.execute(count_sql, sql_values)
+            room_count = txn.fetchone()
+            if room_count:
+                return rows, room_count[0]
+
+            # room_count can be None if there are no rows and OFFSET > 1
+            # Set to 0 instead
+            return rows, 0
+
+        room_tuples, total_rooms = await self.db.runInteraction(
```

Comment:
`total_rooms` isn't (currently) the actual total number of rooms, but rooms left. Though as per above I'd vote we change that.